asciigraf/asciigraf.py

```diff
@@ -105,62 +68,208 @@ def graph_from_ascii(network_string):
+            raise InvalidEdgeError(
+                f"Too many nodes:\nNetwork String:\n"
+                f"{network_string}\nAffected Edge:\n{draw(edge_map)}")
         elif len(edge['nodes']) < 2:
-            raise TooFewNodesOnEdge(edge)
-        else:
-            ascii_graph.add_edge(*edge['nodes'])
-    networkx.set_edge_attributes(ascii_graph, name="length", values={
-        tuple(edge["nodes"]): len(edge["points"])
-        for edge in edges if edge["nodes"]
-    })
+            raise InvalidEdgeError(
+                f"Too few nodes:\nNetwork String:\n"
+                f"{network_string}\nAffected Edge:\n{draw(edge_map)}")
+    ascii_graph = networkx.OrderedGraph()
+    ascii_graph.add_nodes_from(
+        (node, {"position": position}) for position, node in nodes.items()
+    )
+    ascii_graph.add_edges_from(
+        (*edge['nodes'], {"length": len(edge["points"])})
+        for edge in edges
+    )
     networkx.set_edge_attributes(
         ascii_graph, name="label",
         values={
-            tuple(edge_char_to_edge_map[pos]["nodes"]): label
-            for pos, label in line_labels.items()
+            tuple(edge_char_to_edge_map[pos]["nodes"]): label[1:-1]
+            for pos, label in label_char_to_label.items()
+            if pos in edge_char_to_edge_map
         }
     )
     return ascii_graph
 
 
-def char_iter(network_string):
-    return (
+def get_nodes_and_labels(network_string):
+    """ Map the root position of nodes and labels
+        to the node / label text.
+
+        e.g. map_nodes_and_labels("  n1--(label1)--n2  ") -> {
+            Point(2, 0): "n1",
+            Point(6, 0): "(label1)",
+            Point(16, 0): "n2",
+        }
+    """
+    ascii_nodes = list(node_iter(network_string))
+    nodes = OrderedDict()  # of the form {Point -> 'node_name'}
+    labels = OrderedDict()  # of the form {Point -> 'label'}
+    for ascii_label, root_position in ascii_nodes:
+        if ascii_label.startswith("(") and ascii_label.endswith(")"):
+            labels[root_position] = ascii_label
+        else:
+            nodes[root_position] = ascii_label
+    return nodes, labels
+
+
+def get_edge_chars(network_string):
+    """ Map positions in the string to edge chars
+
+        e.g. get_edge_chars("  --|   ") -> {
+            Point(3,0): "-",
+            Point(4,0): "-",
+            Point(5,0): "|",
+        }
+    """
+    return OrderedDict(
         (Point(col, row), char)
         for row, line in enumerate(network_string.split("\n"))
         for col, char in enumerate(line)
+        if char in EDGE_CHARS
+    )
+
+
+def patch_edge_chars_over_labels(labels, edge_chars):
+    """ Adds in edge chars where labels crossed an edge
+
+        e.g.
+
+        ---(horizontal_label)---
+
+                becomes
+
+        ------------------------
+
+        e.g.
+                |                         |
+          (vertical_label)   becomes      |
+                |                         |
```

Comment:
I think this becomes `------|---------`

Reply:
I added a test showing that this case succeeds.